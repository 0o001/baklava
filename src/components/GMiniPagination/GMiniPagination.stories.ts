import { storiesOf } from '@storybook/vue';
import { withKnobs, number } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import GMiniPagination from './GMiniPagination.vue';

const stories = storiesOf('GMiniPagination', module);
stories.addDecorator(withKnobs);

stories
  .addParameters({ component: GMiniPagination })
  .add('base', () => ({
    components: { GMiniPagination },
    props: {
      totalPages: {
        default: number('Total Pages', 3),
      },
    },
    template: `
      <div style="padding: 2rem;">
        <GMiniPagination :totalPages="totalPages" @page="onPage" v-model="currentPage"/>
        <span>Current Page: {{currentPage}}</span>
      </div>
    `,
    data () {
      return {
        currentPage: 1,
      };
    },
    methods: {
      onPage (data) {
        action('Page')(data);
      },
    },
  }));
