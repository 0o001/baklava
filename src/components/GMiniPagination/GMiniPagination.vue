<template>
  <GBox
    :flex="true"
    align="items-center"
    class="g-mini-pageable-wrapper"
  >
    <GIcon
      name="chevron-left"
      class="cursor-pointer navigator prev"
      @click="onPage('prev')"
    />
    <GText
      variant="subtitle-03"
      color="orange-500"
      class="current-page"
    >
      {{ currentPage }}
    </GText>
    <GText
      variant="subtitle-03"
      class="separator"
    >
      /
    </GText>
    <GText
      variant="subtitle-03"
      class="total-pages"
    >
      {{ totalPages }}
    </GText>
    <GIcon
      name="chevron-right"
      class="cursor-pointer navigator next"
      @click="onPage('next')"
    />
  </GBox>
</template>

<script lang="ts">

import GIcon from '../GIcon/GIcon.vue';
import GText from '../GText/GText.vue';
import GBox from '../GBox/GBox.vue';

export default {
  name: 'GMiniPagination',

  components: {
    GIcon,
    GText,
    GBox,
  },

  props: {
    totalPages: {
      type: Number,
      required: true,
    },
    value: {
      type: Number,
      default: null,
    },
  },

  data () {
    return {
      currentPage: 1,
    };
  },

  methods: {
    onPage (direction: String) {
      const { totalPages, value, currentPage } = this;
      let newPage = value || currentPage;

      if (direction === 'prev' && newPage > 1) {
        newPage = newPage - 1;
      } else if (direction === 'next' && newPage < totalPages) {
        newPage = newPage + 1;
      }

      if (newPage !== currentPage) {
        this.currentPage = newPage;

        this.$emit('input', newPage);
        this.$emit('page', { totalPages, page: newPage, direction });
      }
    },
  },
};
</script>

<style lang="scss" scoped>
  .g-mini-pageable-wrapper {
    .navigator {
      cursor: pointer;
    }
    .current-page {
      user-select: none;
      margin-right: 3px;
    }
    .separator {
      user-select: none;
    }
    .total-pages {
      user-select: none;
      margin-left: 3px;
    }
  }
</style>
