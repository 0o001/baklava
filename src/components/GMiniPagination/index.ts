import GMiniPagination from './GMiniPagination.vue';

export default GMiniPagination;
