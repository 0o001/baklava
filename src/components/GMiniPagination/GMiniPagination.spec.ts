import { shallowMount } from '@vue/test-utils';
import GMiniPagination from './';

describe('GMiniPagination', () => {
  let wrapper: any;
  let propsData;

  beforeEach(() => {
    propsData = {
      totalPages: 3,
    };
    wrapper = shallowMount(GMiniPagination, {
      propsData,
    });
  });

  it('should match snapshot on initial', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should use given model value if v-model is used', async () => {
    wrapper.setProps({
      value: 2,
    });
    await wrapper.vm.$nextTick();

    wrapper.vm.onPage('next');

    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().input).not.toBe(undefined);
    expect(wrapper.emitted().input[0][0]).toEqual(3);
    expect(wrapper.find('.current-page').text()).toBe('3');
  });

  it('should emit page event if direction is next and currentPage is less than totalPages', async () => {
    wrapper.vm.onPage('next');

    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().page).not.toBe(undefined);
    expect(wrapper.emitted().page[0][0]).toEqual({
      page: 2,
      direction: 'next',
      totalPages: 3,
    });
    expect(wrapper.find('.current-page').text()).toBe('2');
  });

  it('should emit page event if direction is next and currentPage is greater than 1', async () => {
    wrapper.setData({
      currentPage: 2,
    });
    wrapper.vm.onPage('prev');

    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().page).not.toBe(undefined);
    expect(wrapper.emitted().input).not.toBe(undefined);
    expect(wrapper.emitted().input[0][0]).not.toBe(3);
    expect(wrapper.emitted().page[0][0]).toEqual({
      page: 1,
      direction: 'prev',
      totalPages: 3,
    });
    expect(wrapper.find('.current-page').text()).toBe('1');
  });

  it('should not emit page event if currentPage is equal to totalPages', async () => {
    wrapper.setData({
      currentPage: 3,
    });
    wrapper.vm.onPage('next');

    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().page).toBe(undefined);
    expect(wrapper.find('.current-page').text()).toBe('3');
  });

  it('should not emit page event if currentPage is equal to 1', async () => {
    wrapper.setData({
      currentPage: 1,
    });
    wrapper.vm.onPage('prev');

    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().page).toBe(undefined);
    expect(wrapper.find('.current-page').text()).toBe('1');
  });
});
