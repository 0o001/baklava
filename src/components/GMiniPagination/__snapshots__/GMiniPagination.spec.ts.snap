// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GMiniPagination should match snapshot on initial 1`] = `
<gbox-stub
  align="items-center"
  class="g-mini-pageable-wrapper"
  flex="true"
>
  <gicon-stub
    class="cursor-pointer navigator prev"
    fill="none"
    name="chevron-left"
    size="24px"
    stroke="currentColor"
    strokelinecap="round"
    strokelinejoin="round"
    strokewidth="2"
    tag="i"
  />
   
  <gtext-stub
    class="current-page"
    color="orange-500"
    variant="subtitle-03"
  >
    
    1
  
  </gtext-stub>
   
  <gtext-stub
    class="separator"
    color="main-grey-500"
    variant="subtitle-03"
  >
    
    /
  
  </gtext-stub>
   
  <gtext-stub
    class="total-pages"
    color="main-grey-500"
    variant="subtitle-03"
  >
    
    3
  
  </gtext-stub>
   
  <gicon-stub
    class="cursor-pointer navigator next"
    fill="none"
    name="chevron-right"
    size="24px"
    stroke="currentColor"
    strokelinecap="round"
    strokelinejoin="round"
    strokewidth="2"
    tag="i"
  />
</gbox-stub>
`;
